Crashlytics/Crashlytics/Controllers/FIRCLSMetricKitManager.m

```diff
@@ -0,0 +1,320 @@
+#import "Crashlytics/Crashlytics/Models/FIRCLSExecutionIdentifierModel.h"
+#import "Crashlytics/Crashlytics/Models/FIRCLSInternalReport.h"
+#import "Crashlytics/Crashlytics/Public/FirebaseCrashlytics/FIRCrashlyticsReport.h"
+
+@interface FIRCLSMetricKitManager ()
+
+@property FBLPromise *metricKitDataAvailable;
+@property FIRCLSExistingReportManager *existingReportManager;
+@property FIRCLSFileManager *fileManager;
+@property FIRCLSManagerData *managerData;
+
+@end
+
+@implementation FIRCLSMetricKitManager
+
+- (instancetype)initWithManagerData:(FIRCLSManagerData *)managerData
+              existingReportManager:(FIRCLSExistingReportManager *)existingReportManager
+                        fileManager:(FIRCLSFileManager *)fileManager {
+  _existingReportManager = existingReportManager;
+  _fileManager = fileManager;
+  _managerData = managerData;
+  return self;
+}
+
+/*
+ * Registers the MetricKit manager to receive MetricKit reports by adding self to the
+ * MXMetricManager subscribers. Also initializes the promise that we'll use to ensure that any
+ * MetricKit report files are included in Crashylytics fatal reports. If no crash occurred on the
+ * last run of the app, this promise is immediately resolved so that the upload of any nonfatal
+ * events can proceed.
+ */
+- (void)registerMetricKitManager {
+  [[MXMetricManager sharedManager] addSubscriber:self];
+  self.metricKitDataAvailable = [FBLPromise pendingPromise];
+
+  // If there was no crash on the last run of the app, then we aren't expecting a MetricKit
+  // diagnostic report and should resolve the promise immediately.
+  if (![self.fileManager didCrashOnPreviousExecution] && ![self.fileManager didCrashPreviously]) {
+    @synchronized(self) {
+      [self.metricKitDataAvailable fulfill:nil];
+    }
+  }
+}
+
+/*
+ * This method receives diagnostic payloads from MetricKit whenever a fatal or nonfatal MetricKit
+ * event occurs. If a fatal event, this method will be called when the app restarts. Since we're
+ * including a MetricKit report file in the Crashlytics report to be sent to the backend, we need
+ * to make sure that we process the payloads and write the included information to file before
+ * the report is sent up. If this method is called due to a nonfatal event, it will be called
+ * immediately after the event. Since we send nonfatal events on the next run of the app, we can
+ * write out the information but won't need to resolve the promise.
+ */
+- (void)didReceiveDiagnosticPayloads:(NSArray<MXDiagnosticPayload *> *)payloads {
+  NSUInteger payloadCount = [payloads count];
+  FIRCLSInfoLog(@"[Crashlytics:Crash] received %lu diagnostic payloads from MetricKit",
+                payloadCount);
+
+  if ([payloads count] == 0) {
+    return;
+  }
+
+  for (MXDiagnosticPayload *diagnosticPayload in payloads) {
+    if (!diagnosticPayload) {
+      return;
+    }
+
+    // TODO: Time stamp information is only available in begin and end time periods. Hopefully this
```

Comment:
Can we move everything from this line to 224 to a new method? We return early from this method in some cases, which means we don't resolve the promise and don't upload reports. We can have it return success or failure instead.